File: bin/NJ.py
#!/usr/bin/env python3
import sys

def read():
  f = open(sys.argv[1], 'r')
  file = f.read().strip()
  f.close
  return file

def write(thing):
  f = open("out.txt", 'w')
  f.write(thing)
  f.close()

def adjust(out, node):
  score = 0
  path = []
  while len(out[node]) > 1:
    path.append(node)
    for index in out[node]:
      if index[1] not in path:
        score += index[2]
        path.append(index[1])
        break
    node = path[-1]
  return score

def countNodes(out, node):
  count = 0
  path = [node]
  tempPath = []
  while list(tempPath) != list(path):
    tempPath = list(path)
    for index in path:
      for line in out[index]:
        if line[1] not in path:
          path.append(line[1])
  for index in path:
    if len(out[index]) <= 1:
      count += 1
  return count

def branch(matrix, tree, count):
  d = []
  for h in range(0, len(matrix)):
    d.append([])
    for v in range(0, len(matrix)):
      d[h].append(0)
      if h != v:
        d[h][v] = (len(matrix) - 2) * matrix[h][v] - sum(matrix[h]) - sum(matrix[v])

  best = [d[0][1], 0, 1]
  for i in range(0, len(d)):
    for j in range(i + 1, len(d)):
      if i != j and d[i][j] < best[0]:
        best = [d[i][j], i, j]
  i = best[1]
  j = best[2]

  for k in range(0, len(d[i])):
    if d[i][k] == best[0] and k != j:
      i = k
      break

  for index in d:
    print("  ",index)

  if i == j and j == 0:
    print("omergerrrrd")
    quit()
  return [matrix, i, j, [tree[i], best[0]/2, tree[j], best[0]/2, count]]

def main():
  file = read().split(">")[1:]
  nLeaves = len(file)
  leaves = []
  for i in range(nLeaves):
    file[i] = file[i].split("\n")
    while '' in file[i]:
      file[i].remove('')
    leaves.append(file[i].pop(0))
    file[i] = "".join(file[i])

  print(leaves)
  matrix = []
  for i in range(len(file)):
    print(i,"of",len(file))
    matrix.append([])
    for j in range(len(file)):
      if i == j:
        matrix[i].append(0)
      elif len(matrix) > j and len(matrix[j]) > i:
        print(i,j)
        matrix[i].append(matrix[j][i])
      else:
        total = 0
        for k in range(len(file[i])):
          if file[i][k] != file[j][k]:
            total += 1
        matrix[i].append(total)


  count = len(matrix) -1
  tree = []
  for i in range(0, len(matrix)):
    tree.append(i)
  print(tree)
  out = {}

  for index in matrix:
    print(" ",index)

  while len(matrix) > 2:
    count += 1
    temp = branch(matrix, tree, count)
    matrix = temp[0]
    i = temp[1]
    j = temp[2]

    if temp[3][0] not in out:
      out[temp[3][0]] = []
    if temp[3][2] not in out:
      out[temp[3][2]] = []
    if temp[3][4] not in out:
      out[temp[3][4]] = []

    delta = sum(matrix[i])
    for k in range(0, len(matrix)):
      delta -= matrix[k][j]
    delta /= len(matrix) - 2
    iLimb = (matrix[i][j] + delta) / 2
    jLimb = (matrix[i][j] - delta) / 2

    print(tree, i, j, iLimb, jLimb)

    for h in range(0, len(matrix)):
      if h != i:
        print("====>", i, j, matrix[h][i], matrix[h][j], matrix[i][j])
        matrix[h][i] = (matrix[h][i] + matrix[h][j] - matrix[i][j])/2

    for v in range(0, len(matrix)):
      matrix[i][v] = matrix[v][i]

    for v in range(0, len(matrix)):
      matrix[v].pop(j)

    matrix.pop(j)

    for index in matrix:
      print(" ",index)

    out[temp[3][0]].append([temp[3][0], temp[3][4], iLimb])
    out[temp[3][2]].append([temp[3][2], temp[3][4], jLimb])
    out[temp[3][4]].append([temp[3][4], temp[3][0], iLimb])
    out[temp[3][4]].append([temp[3][4], temp[3][2], jLimb])

    tree[i] = count
    tree.pop(j)

  if tree[0] not in out:
    out[tree[0]] = []
  if tree[1] not in out:
    out[tree[1]] = []
  if count not in out:
    out[count] = []
  out[tree[0]].append([tree[0], tree[1], matrix[0][1]])
  out[tree[1]].append([tree[1], tree[0], matrix[0][1]])

  for index in sorted(out.keys()):
    for value in out[index]:
      value[2] = str(round(value[2], 3)) + "000"
      value[2] = value[2].split(".")[0] + "." + value[2].split(".")[1][:3]
      string = str(value[0]) + "->" + str(value[1]) + ":" + value[2]
      print(string)

#============================================================( run main )
###### runMain
if __name__ == "__main__":
  main()

File: go/NJ.go
package main

import "fmt"
import "io/ioutil"
import "os"
import "strings"
import "strconv"
import "sync"
import "math"

func printHelp() {
  fmt.Println("\nconverts a FASTA file to neighbor joined NWK tree")
  fmt.Println("usage: NJ [-h] FASTA")
  fmt.Println("\tFASTA:\tfasta file, ie. 'bestsnp.fasta'")
  fmt.Println("example:\n\t./NJ bestsnp.fasta\n")
  os.Exit(0)
}

func sliceSum(s1 []int) (sum int) {
  for i := 0; i < len(s1); i++ {
    sum += s1[i]
  }
  return
}

func hamming(s1 string, s2 string) (distance int) {
  // index by code point, not byte
  r1 := []rune(s1)
  r2 := []rune(s2)

  for i, v := range r1 {
    if r2[i] != v {
      distance += 1
    }
  }
  return
}

func indexList(a int, list []int) bool {
    for _, b := range list {
        if b == a {
            return true
        }
    }
    return false
}

func branch(matrix [][]int, tempTree []int, count int) (b1 int, b2 int){

  b1 = 0
  b2 = 1
  best := (len(matrix) - 2) * matrix[b1][b2] - sliceSum(matrix[b1]) - sliceSum(matrix[b2])

  for i := 0; i < len(matrix); i++ {
    for j := i+1; j < len(matrix); j++ {
      if (len(matrix) - 2) * matrix[i][j] - sliceSum(matrix[i]) - sliceSum(matrix[j]) < best {
        b1 = i
        b2 = j
        best = (len(matrix) - 2) * matrix[b1][b2] - sliceSum(matrix[b1]) - sliceSum(matrix[b2])
      }
    }
  }
  return
}

func getBest(tree map[int]map[int]float64, tempScore float64, tempPath []int, bestScore []float64, bestPath []int) ([]int){
  if tempScore > bestScore[0] {
    bestScore[0] = tempScore
    bestPath = bestPath[:0]
    for i := 0; i < len(tempPath); i ++ {
      bestPath = append(bestPath, tempPath[i])
    }
  }
  for key, _ := range tree[tempPath[len(tempPath) - 1]] {
    if !indexList(key, tempPath) {
      bestPath = getBest(tree, tempScore + tree[tempPath[len(tempPath) - 1]][key], append(tempPath, key), bestScore, bestPath)
    }
  }
  return bestPath
}

func getNewick(names []string, tree map[int]map[int]float64, path []int, score float64) string {
  if path[len(path) - 1] < len(names) && path[len(path) - 1] >= 0 {
    return names[path[len(path) - 1]] + ":" + strconv.FormatFloat(math.Abs(score), 'f', -1, 64)
  } else if score == 0 {
    thing := []string{}
    for index := range tree[path[len(path) - 1]]{
      if ! indexList(index, path) {
  
        thing = append(thing, getNewick(names, tree, append(path, index), tree[path[len(path) - 1]][index]))
      }
    }
    return strings.Join(thing, ",")
  } else {
    thing := []string{}
    for index := range tree[path[len(path) - 1]]{
      if ! indexList(index, path) {
        thing = append(thing, getNewick(names, tree, append(path, index), tree[path[len(path) - 1]][index]))
      }
    }
    return "(" + strings.Join(thing, ",") + "):" + strconv.FormatFloat(math.Abs(score), 'f', -1, 64)
  }
}

func main() {
  // read fasta
  var fileName string
  if len(os.Args) <= 1 {
    printHelp()
  } else {
    for i := range os.Args[:1] {
      if os.Args[i] == string("-h") {
        printHelp()
      }
    }
    fileName = os.Args[1]
  }
  f, err := ioutil.ReadFile(fileName)
  if err != nil {
    fmt.Println()
    fmt.Println(err)
    printHelp()
  }
  DIR := strings.Join(strings.Split(fileName, "/")[:len(strings.Split(fileName, "/"))-1], "/")
  fileName = strings.Split(fileName, "/")[len(strings.Split(fileName, "/")) - 1]
  fileName = strings.Split(fileName, ".fasta")[0]
  newick := DIR + "/" + fileName + ".nwk"
  fasta := strings.Split(string(f), ">")[1:]
  names := []string{}

  // hash fasta
  for i := 0; i < len(fasta); i++ {
    names = append(names, strings.Split(fasta[i], "\n")[0])
    fasta[i] = strings.Join(strings.Split(fasta[i], "\n")[1:], "")
  }

  // mk matrix
  matrix := make([][]int, len(names))
  for i := 0; i < len(names); i++ {
    matrix[i] = make([]int, len(names))
    matrix[i][i] = 0
  }

  // build matrix
  var wg sync.WaitGroup
  for i := 0; i < len(names); i++ {
    for j := 0; j < len(names); j++ {
      wg.Add(1)
      go func(i int, j int) {
        defer wg.Done()
        matrix[i][j] = hamming(fasta[i], fasta[j])
      }(i, j)
    }
    //fmt.Println(i, len(names))
  }
  wg.Wait()

  // mirror matrix
  for i := 0; i < len(names); i++ {
    for j := i+1; j < len(names); j++ {
      wg.Add(1)
      go func(i int, j int) {
        defer wg.Done()
        matrix[j][i] = matrix[i][j]
      }(i, j)
    }
  }
  wg.Wait()

  // prep tempTree
  count := len(names) - 1
  tempTree := []int{}
  var tree map[int]map[int]float64
  tree = make(map[int]map[int]float64)
  for i := 0; i < len(names); i++ {
    tempTree = append(tempTree, i)
    tree[i] = make(map[int]float64)
  }

  //out := []string{}

  // reduce matrix / build tree
  for i := 0; i < len(names) - 2; i++ {
    count ++
    h, v := branch(matrix, tempTree, count)

    delta := 0
    for index := range matrix {
      delta += matrix[h][index]
    }
    for index := range matrix {
      delta -= matrix[index][v]
    }
    delta /= (len(matrix) - 2)

    // distance between tempTree[h] and count
    hLimb := (float64(matrix[h][v]) + float64(delta)) / 2

    // distance between tempTree[v] and count
    vLimb := (float64(matrix[h][v]) - float64(delta)) / 2

    // update matrix COLUMN h
    for index := range matrix {
      if index != h {
        matrix[index][h] = (matrix[index][h] + matrix[index][v] - matrix[h][v]) / 2
      }
    }

    // update matrix ROW h
    for index := range matrix {
      matrix[h][index] = matrix[index][h]
    }

    // remove COLUMN and ROW v
    for index := range matrix {
      matrix[index] = append(matrix[index][:v], matrix[index][v+1:]...)
    }
    matrix = append(matrix[:v], matrix[v+1:]...)

    // update h, remove v from tempTree
    tree[count] = make(map[int]float64)
    tree[count][tempTree[h]] = hLimb
    tree[count][tempTree[v]] = vLimb
    tree[tempTree[h]][count] = hLimb
    tree[tempTree[v]][count] = vLimb
    

    tempTree[h] = count
    tempTree = append(tempTree[:v], tempTree[v+1:]...)

    // getbest
  }
  tree[count + 1] = make(map[int]float64)
  //tree[count + 1][tempTree[0]] = float64(matrix[0][1]) / 2
  //tree[count + 1][tempTree[1]] = float64(matrix[0][1]) / 2
  //tree[tempTree[0]][count + 1] = float64(matrix[0][1]) / 2
  //tree[tempTree[1]][count + 1] = float64(matrix[0][1]) / 2
  tree[tempTree[0]][tempTree[1]] = float64(matrix[0][1])
  tree[tempTree[1]][tempTree[0]] = float64(matrix[0][1])

  // get nodes with longest path
  bestScore := []float64{0}
  bestPath := []int{}
  for i := 0; i < len(names); i++ {

    bestPath = getBest(tree, 0, []int{i}, bestScore, bestPath)
  }

  // get midpoint location and root at midpoint
  bestScore[0] -= bestScore[0] / 2
  tree[-1] = make(map[int]float64)
  for i := 0; i < len(bestPath) - 1; i++ {
    if bestScore[0] - tree[bestPath[i]][bestPath[i + 1]] < 0 {
      tree[-1][bestPath[i]] = bestScore[0]
      tree[-1][bestPath[i + 1]] = -(bestScore[0] - tree[bestPath[i]][bestPath[i + 1]])
      delete(tree[bestPath[i]], bestPath[i + 1])
      delete(tree[bestPath[i + 1]], bestPath[i])
      break
    }
    bestScore[0] -= tree[bestPath[i]][bestPath[i + 1]]
  }

  // print nwk
  path := []int{-1}
  out := "(" + getNewick(names, tree, path, 0) + "):0;"

  outFile, err := os.Create(newick)
  if err != nil {
    fmt.Println(err)
    return
  }
  _, err = outFile.WriteString(out)
  if err != nil {
    fmt.Println(err)
    outFile.Close()
    return
  }

  os.Exit(0)

}
















